src/native-shim.js

```diff
@@ -18,15 +18,125 @@
  * shim must only be loaded in browsers that do.
  */
 (() => {
-  let origHTMLElement = HTMLElement;
-  // TODO(justinfagnani): Tests!!
+  'use strict';
+
+  const NativeHTMLElement = HTMLElement;
+  const nativeDefine = window.customElements.define;
+  const nativeGet = window.customElements.get;
+
+  /**
+   * Map of user-provided constructors to tag names.
+   *
+   * @type {Map<Function, string>}
+   */
+  const tagnameByConstructor = new Map();
+
+  /**
+   * Map of tag anmes to user-provided constructors.
+   *
+   * @type {Map<string, Function>}
+   */
+  const constructorByTagname = new Map();
+
+
+  // /**
+  //  * Whether the constructors are being called by a browser process, ie parsing
+  //  * or createElement.
+  //  */
+  // let browserConstruction = false;
+  //
+  // /**
+  //  * Whether the constructors are being called by a user-space process, ie
+  //  * calling an element constructor.
+  //  */
+  // let userConstruction = false;
+
   window.HTMLElement = function() {
-    // prefer new.target for elements that call super() constructors or
-    // Reflect.construct directly
-    let newTarget = new.target || this.constructor;
-    return Reflect.construct(origHTMLElement, [], newTarget);
-  }
-  HTMLElement.prototype = Object.create(origHTMLElement.prototype, {
-    constructor: {value: HTMLElement, configurable: true, writable: true},
-  });
+    if (!HTMLElement.browserConstruction) {
+      const tagname = tagnameByConstructor.get(this.constructor);
+      const fakeClass = nativeGet.call(window.customElements, tagname);
+
+      // Make sure that the fake constructor doesn't call back to this constructor
+      HTMLElement.userConstruction = true;
+      const instance = new (fakeClass)();
+      return instance;
+    }
+    // Else do nothing. This will be reached by ES5-style classes doing
+    // HTMLElement.call() during initialization
+    HTMLElement.browserConstruction = false;
+  };
+
+  HTMLElement.prototype = Object.create(NativeHTMLElement.prototype);
+  HTMLElement.prototype.constructor = HTMLElement;
+  HTMLElement.nativeClass = NativeHTMLElement;
+
+
+  /**
+   * Whether the constructors are being called by a browser process, ie parsing
+   * or createElement.
+   */
+  HTMLElement.browserConstruction = false;
+
+  /**
+   * Whether the constructors are being called by a user-space process, ie
+   * calling an element constructor.
+   */
+  HTMLElement.userConstruction = false;
+
+  window.customElements.define = (tagname, elementClass) => {
```

Comment:
It'd be nice to check that elementClass is an ES5-style class and throw if not. This approach requires that all constructors in a chain are ES5 - that the bottom-most constructor can be called with `Function.call()`